From 717719db35310c7f8b1398d335a42d76a143f0f1 Mon Sep 17 00:00:00 2001
From: Donald Sharp <sharpd@nvidia.com>
Date: Wed, 22 Oct 2025 13:40:33 -0400
Subject: [PATCH 2/7] zebra: Allow bonds to refigure speed

When a member of a bond goes down, notice
that this has happened and schedule a interface
speed update.  This will allow the interface
speed to stay accurate to some extent.

Signed-off-by: Donald Sharp <sharpd@nvidia.com>
---
 zebra/interface.c | 23 +++++++++++++++++++++++
 1 file changed, 23 insertions(+)

diff --git a/zebra/interface.c b/zebra/interface.c
index e1fd092532..100e6c10fc 100644
--- a/zebra/interface.c
+++ b/zebra/interface.c
@@ -911,6 +911,25 @@ static void if_down_del_nbr_connected(struct interface *ifp)
 	}
 }
 
+static void if_handle_bond_speed_change(struct interface *ifp)
+{
+	struct zebra_if *zif = ifp->info;
+	struct zebra_l2info_bondslave *part_of_bond;
+
+	if (!IS_ZEBRA_IF_BOND_SLAVE(ifp))
+		return;
+
+	part_of_bond = &zif->bondslave_info;
+
+	if (part_of_bond->bond_if) {
+		zif = part_of_bond->bond_if->info;
+
+		if (!event_is_scheduled(zif->speed_update))
+			event_add_timer(zrouter.master, if_zebra_speed_update, part_of_bond->bond_if, 1,
+					&zif->speed_update);
+	}
+}
+
 /* Interface is up. */
 void if_up(struct interface *ifp, bool install_connected)
 {
@@ -975,6 +994,8 @@ void if_up(struct interface *ifp, bool install_connected)
 
 	if_addr_wakeup(ifp);
 
+	if_handle_bond_speed_change(ifp);
+
 	rib_update_handle_vrf_all(RIB_UPDATE_KERNEL, ZEBRA_ROUTE_KERNEL);
 }
 
@@ -1027,6 +1048,8 @@ void if_down(struct interface *ifp)
 	/* Delete all neighbor addresses learnt through IPv6 RA */
 	if_down_del_nbr_connected(ifp);
 
+	if_handle_bond_speed_change(ifp);
+
 	rib_update_handle_vrf_all(RIB_UPDATE_INTERFACE_DOWN, ZEBRA_ROUTE_KERNEL);
 }
 
-- 
2.43.0

